In [14]:
import pandas as pd


In [15]:
# load raw train data

df = pd.read_csv('https://raw.githubusercontent.com/akwasianing/train_project/refs/heads/main/data_raw/raw_train_data.csv')
df.head()


Unnamed: 0                              Name    Location  Year  \
0           1  Hyundai Creta 1.6 CRDi SX Option        Pune  2015   
1           2                      Honda Jazz V     Chennai  2011   
2           3                 Maruti Ertiga VDI     Chennai  2012   
3           4   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013   
4           6            Nissan Micra Diesel XV      Jaipur  2013   

   Kilometers_Driven Fuel_Type Transmission Owner_Type     Mileage   Engine  \
0              41000    Diesel       Manual      First  19.67 kmpl  1582 CC   
1              46000    Petrol       Manual      First    13 km/kg  1199 CC   
2              87000    Diesel       Manual      First  20.77 kmpl  1248 CC   
3              40670    Diesel    Automatic     Second   15.2 kmpl  1968 CC   
4              86999    Diesel       Manual      First  23.08 kmpl  1461 CC   

       Power  Seats  New_Price  Price  
0  126.2 bhp    5.0        NaN  12.50  
1   88.7 bhp    5.0  8.61 Lakh   4.50  
2  88.76 bhp    7.0        NaN   6.00  
3  140.8 bhp    5.0        NaN  17.74  
4   63.1 bhp    5.0        NaN   3.50

In [16]:
#  drop the unnamed column as it is not a relevant feature
df_cleaned = df.drop(columns=['Unnamed: 0'])
df_cleaned.head()

Name    Location  Year  Kilometers_Driven  \
0  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
1                      Honda Jazz V     Chennai  2011              46000   
2                 Maruti Ertiga VDI     Chennai  2012              87000   
3   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   
4            Nissan Micra Diesel XV      Jaipur  2013              86999   

  Fuel_Type Transmission Owner_Type     Mileage   Engine      Power  Seats  \
0    Diesel       Manual      First  19.67 kmpl  1582 CC  126.2 bhp    5.0   
1    Petrol       Manual      First    13 km/kg  1199 CC   88.7 bhp    5.0   
2    Diesel       Manual      First  20.77 kmpl  1248 CC  88.76 bhp    7.0   
3    Diesel    Automatic     Second   15.2 kmpl  1968 CC  140.8 bhp    5.0   
4    Diesel       Manual      First  23.08 kmpl  1461 CC   63.1 bhp    5.0   

   New_Price  Price  
0        NaN  12.50  
1  8.61 Lakh   4.50  
2        NaN   6.00  
3        NaN  17.74  
4        NaN   3.50

In [17]:
df_cleaned.shape

(5847, 13)

In [18]:
df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5847 entries, 0 to 5846
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               5847 non-null   object 
 1   Location           5847 non-null   object 
 2   Year               5847 non-null   int64  
 3   Kilometers_Driven  5847 non-null   int64  
 4   Fuel_Type          5847 non-null   object 
 5   Transmission       5847 non-null   object 
 6   Owner_Type         5847 non-null   object 
 7   Mileage            5845 non-null   object 
 8   Engine             5811 non-null   object 
 9   Power              5811 non-null   object 
 10  Seats              5809 non-null   float64
 11  New_Price          815 non-null    object 
 12  Price              5847 non-null   float64
dtypes: float64(2), int64(2), object(9)
memory usage: 594.0+ KB


In [19]:
# extract numeric values from attributes (Mileage, Engine, Power, New_Price) and remove units

df_cleaned['Mileage'] = df_cleaned['Mileage'].str.extract(r'(\d+\.?\d*)').astype(float)

df_cleaned['Engine'] = df_cleaned['Engine'].str.extract(r'(\d+\.?\d*)').astype(float)

df_cleaned['Power'] = df_cleaned['Power'].str.extract(r'(\d+\.?\d*)').astype(float)

df_cleaned['New_Price'] = df_cleaned['New_Price'].str.extract(r'(\d+\.?\d*)').astype(float)


In [20]:
# Add units to column names for clarity

column_units = {
    'Mileage': 'Mileage (kmpl)',
    'Engine': 'Engine (cc)',
    'Power': 'Power (bhp)',
    'New_Price': 'New_Price (Lakh)'
}

# Rename the columns
df_cleaned.rename(columns=column_units, inplace=True)

In [21]:
df_cleaned.head(10)

Name    Location  Year  Kilometers_Driven  \
0     Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
1                         Honda Jazz V     Chennai  2011              46000   
2                    Maruti Ertiga VDI     Chennai  2012              87000   
3      Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   
4               Nissan Micra Diesel XV      Jaipur  2013              86999   
5    Toyota Innova Crysta 2.8 GX AT 8S      Mumbai  2016              36000   
6  Volkswagen Vento Diesel Comfortline        Pune  2013              64430   
7       Tata Indica Vista Quadrajet LS     Chennai  2012              65932   
8                     Maruti Ciaz Zeta       Kochi  2018              25692   
9          Honda City 1.5 V AT Sunroof     Kolkata  2012              60000   

  Fuel_Type Transmission Owner_Type  Mileage (kmpl)  Engine (cc)  Power (bhp)  \
0    Diesel       Manual      First           19.67       1582.0       126.20   
1    Petrol       Manual      First           13.00       1199.0        88.70   
2    Diesel       Manual      First           20.77       1248.0        88.76   
3    Diesel    Automatic     Second           15.20       1968.0       140.80   
4    Diesel       Manual      First           23.08       1461.0        63.10   
5    Diesel    Automatic      First           11.36       2755.0       171.50   
6    Diesel       Manual      First           20.54       1598.0       103.60   
7    Diesel       Manual     Second           22.30       1248.0        74.00   
8    Petrol       Manual      First           21.56       1462.0       103.25   
9    Petrol    Automatic      First           16.80       1497.0       116.30   

   Seats  New_Price (Lakh)  Price  
0    5.0               NaN  12.50  
1    5.0              8.61   4.50  
2    7.0               NaN   6.00  
3    5.0               NaN  17.74  
4    5.0               NaN   3.50  
5    8.0             21.00  17.50  
6    5.0               NaN   5.20  
7    5.0               NaN   1.95  
8    5.0             10.65   9.95  
9    5.0               NaN   4.49

In [22]:
# show info after removing units from values

df_cleaned.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5847 entries, 0 to 5846
Data columns (total 13 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Name               5847 non-null   object 
 1   Location           5847 non-null   object 
 2   Year               5847 non-null   int64  
 3   Kilometers_Driven  5847 non-null   int64  
 4   Fuel_Type          5847 non-null   object 
 5   Transmission       5847 non-null   object 
 6   Owner_Type         5847 non-null   object 
 7   Mileage (kmpl)     5845 non-null   float64
 8   Engine (cc)        5811 non-null   float64
 9   Power (bhp)        5811 non-null   float64
 10  Seats              5809 non-null   float64
 11  New_Price (Lakh)   815 non-null    float64
 12  Price              5847 non-null   float64
dtypes: float64(6), int64(2), object(5)
memory usage: 594.0+ KB


In [23]:
# identify missing values in all columns

df_cleaned.isna().sum()

Name                    0
Location                0
Year                    0
Kilometers_Driven       0
Fuel_Type               0
Transmission            0
Owner_Type              0
Mileage (kmpl)          2
Engine (cc)            36
Power (bhp)            36
Seats                  38
New_Price (Lakh)     5032
Price                   0
dtype: int64

In [24]:
# Mileage column - drop rows containing missing values

df_cleaned = df_cleaned.dropna(subset=['Mileage (kmpl)'])

For mileage column since only 2 out of 5847 rows have missing values, these two rows can be dropped. (negligible impact)

In [25]:
# Imputing missing values in the columns (Engine, Power, Seats)
# Calculate skewness to decide imputation method

df_cleaned[['Engine (cc)', 'Power (bhp)', 'Seats']].skew()

Engine (cc)    1.412804
Power (bhp)    1.915841
Seats          1.911640
dtype: float64

Since skewness values for the columns (Engine, Power, Seats) are positive, indicating right-skewed distributions, the missing values in these columns are replaced with their respective medians, as the median is robust to outliers and better suited for skewed data


In [26]:
# Calculate the median of each specified column

medians = df_cleaned[['Engine (cc)', 'Power (bhp)', 'Seats']].median()
print(medians)

Engine (cc)    1497.0
Power (bhp)      98.6
Seats             5.0
dtype: float64


In [27]:
# fill in the missing values with the calculated medians

df_cleaned.loc[:, ['Engine (cc)', 'Power (bhp)', 'Seats']] = df_cleaned.loc[:, ['Engine (cc)', 'Power (bhp)', 'Seats']].fillna(medians)

In [28]:
# drop New_Price column

df_cleaned = df_cleaned.drop(['New_Price (Lakh)'], axis=1)

New_Price column is dropped due to a high percentage of missing values (~86%). With such a significant portion of data missing, retaining the column would likely reduce the quality and interpretability of the analysis. Imputation would rather be unreliable and potentially introduce bias that is why the New_Price column is dropped.

In [29]:
# Verify if missing values for the columns (Engine, Power, Seats) are filled with respective medians; the missing values in the Mileage is dropped and the entire New_Price column is dropped.
df_cleaned.isna().sum()

Name                 0
Location             0
Year                 0
Kilometers_Driven    0
Fuel_Type            0
Transmission         0
Owner_Type           0
Mileage (kmpl)       0
Engine (cc)          0
Power (bhp)          0
Seats                0
Price                0
dtype: int64

In [30]:
# make a copy of df_cleaned data

df_encoded = df_cleaned.copy()


In [31]:
# convert 'Fuel_Type' and 'Transmission' into numerical one-hot encoded columns
# assuming using models not sensitive to multicollinearity such as decision trees

df_encoded = pd.get_dummies(df_cleaned, columns=['Fuel_Type', 'Transmission'], dtype=int)


In [34]:
df_encoded.head()

Name    Location  Year  Kilometers_Driven  \
0  Hyundai Creta 1.6 CRDi SX Option        Pune  2015              41000   
1                      Honda Jazz V     Chennai  2011              46000   
2                 Maruti Ertiga VDI     Chennai  2012              87000   
3   Audi A4 New 2.0 TDI Multitronic  Coimbatore  2013              40670   
4            Nissan Micra Diesel XV      Jaipur  2013              86999   

  Owner_Type  Mileage (kmpl)  Engine (cc)  Power (bhp)  Seats  Price  \
0      First           19.67       1582.0       126.20    5.0  12.50   
1      First           13.00       1199.0        88.70    5.0   4.50   
2      First           20.77       1248.0        88.76    7.0   6.00   
3     Second           15.20       1968.0       140.80    5.0  17.74   
4      First           23.08       1461.0        63.10    5.0   3.50   

   Fuel_Type_Diesel  Fuel_Type_Petrol  Transmission_Automatic  \
0                 1                 0                       0   
1                 0                 1                       0   
2                 1                 0                       0   
3                 1                 0                       1   
4                 1                 0                       0   

   Transmission_Manual  
0                    1  
1                    1  
2                    1  
3                    0  
4                    1

In [35]:
# Save the cleaned DataFrame to a CSV file

df_encoded.to_csv('cleaned_train_data.csv', index=False)
